### DISCUSSION TITLE
ITMS-90078: Missing Push Notification Entitlement

### DISCUSSION THREAD

**Discussion by raenk** [codemagic-ci-cd/codemagic-docs#1955] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

I let codemagic generate certificates and identities automatically. I checked these have Push Notifications capabilities at developer console.

Since I develop in Linux, I don't use XCode, so I manually added the following code to my Info.plist file:
```
	<key>UIBackgroundModes</key>
    	<array>
        	<string>fetch</string>
        	<string>remote-notification</string>
    	</array>
```
Built and publish process goes well. However, right after the build is published I get an email from Apple about 'Missing Push Notification Entitlement'.

And the App does not run in TestFlight. Shows a blank screen.

### Expected results

No complains about 'Missing Push Notification Entitlement' and a working App.

### Actual results

Apple notifies me about 'Missing Push Notification Entitlement' and the App does not do anything in TestFlight (blank screen).

### Build id (optional)

64d2f5794f658a2519a73e20

**Comment by kulakarla** [comment, marked as answer]
Hello @raenk

I reproduced it and got the emails too. Yes, it should be rather simple fix in Xcode, but as you are working with Linux, try to do it this way.
First, remove your above added info.plist value.

- Start a build and enable VNC and SSH access
- Connect to the machine
- Open the file `Users/builder/clone/PROJECTNAME/ios/FILENAME.xcodeproj` with Xcode
- On the left side, click on `Runner`, then on the right side `Signing and Capabilities`
- There is a little `+ Capability ` sign/button beneath it. Click it, search for Push Notifications and add the capability. This makes some changes in the iOS specific files.
- After that, close Xcode and open the terminal and navigate to `/Users/builder/clone`. (your repository)
- From there `git` add, commit & push the changes to your repository. Check out these [docs](https://docs.codemagic.io/troubleshooting/accessing-builder-machine-via-ssh/#making-configuration-changes) on how to authenticate yourself to your Git provider.

Give the build another go, and see if you still have issues. If you still run into any issues, let me know.

**Reply by raenk** [reply]
I'll give it a go once i get back from vacation next week and let you know.
Thanks for your help!

**Reply by giuseppemacc** [reply]
Is there a way to solve this problem just by using Codemagic? (because I don't have a mac with xcode)

**Comment by raenk** [comment]
@kulakarla It worked! Although I couldn't push to the repo directly (auth problems), I managed to upload the modified ios folder from the VM to dropbox and merge it into my project locally.

Many thanks again!

**Reply by kulakarla** [reply]
@raenk 

Glad that it worked! If any issues arise, come ask us again ;) Until then, happy building!
